### ISSUE TITLE
Can't merge edges

### ISSUE THREAD

**Issue by zhangwenxuan322** [neo4j-contrib/neo4j-spark-connector#135]
I can create nodes but can't merge edges with Neo4jDataframe.mergeEdgeList(). Here is my code:
```
    val nodes = converter.dataTransfer()
    Neo4jDataFrame.createNodes(spark.sparkContext, nodes,
      ("shebei", nodes.columns.toSeq), partitions = 16, merge = true)

    // edges have two columns: start_id,end_id
    // in neo4j that property is ID
    val edges = converter.createEdges(nodes)
    val rename = Map("start_id" -> "ID", "end_id" -> "ID")
    Neo4jDataFrame.mergeEdgeList(spark.sparkContext, edges,
      ("shebei", Seq("start_id")), ("ACTED_WITH", Seq.empty), ("shebei", Seq("end_id")),
      rename, partitions = 16)
```
The program was just stuck at mergeEdgeList() and couldn't merge edges in neo4j.
Thanks for your help!

**Comment by jexp** [comment]
Do you have an index/constraint on: shebei(ID)  ?
Do you use `ID` in nodes.columns ?

**Comment by zhangwenxuan322** [comment]
1. shebei(ID) has index
2. Yes, I use ID in nodes.columns

**Comment by jexp** [comment]
How big is your dataframe, so that we can try to reproduce.
Is anything in the neo4j logs that could indicate an issue?

**Comment by zhangwenxuan322** [comment]
1. nodes df has 1,070,000 lines of data, creating these nodes costs 3 minute.
2. edges df has 880,000 lines of data, none of edges were merged.
3. this error keeps occurring in the log file:
```
ERROR [o.n.b.t.TransportSelectionHandler] Fatal error occurred when initialising pipeline: [id: 0x3f53be4a, L:/172.16.221.100:7687 ! R:/172.16.224.19:1940] javax.net.ssl.SSLException: Received fatal alert: certificate_unknown
io.netty.handler.codec.DecoderException: javax.net.ssl.SSLException: Received fatal alert: certificate_unknown
```
While creating nodes this error happens too, but nodes can be created.

**Comment by jexp** [comment]
Do you have certificates on your neo4j server?
Can you try the neo4j.encryption=false parameter?

**Comment by conker84** [comment]
@zhangwenxuan322 did you try with Michael suggestion?

**Comment by zhangwenxuan322** [comment]
1. I have certificates on neo4j server.
2. I'm gonna try it on the environment tomorrow.

**Comment by zhangwenxuan322** [comment]
I set the config `spark.conf.set("spark.neo4j.encryption", "false")`. 
I got the following error:
```
org.neo4j.driver.exceptions.TransientException: Can't acquire exclusive lock, because it would have caused a deadlock: ForsetiClient[5] can't acquire ExclusiveLock{owner=ForsetiClient[12]} on NODE(1294719), because holders of that lock are waiting for ForsetiClient[5].
Waitlist: ExclusiveLock[Client[12] waits for [5]]
```
Merging edges still failed.

**Comment by conker84** [comment]
@zhangwenxuan322 can you try with:
```
Neo4jDataFrame.mergeEdgeList(spark.sparkContext, edges,
      ("shebei", Seq("start_id")), ("ACTED_WITH", Seq.empty), ("shebei", Seq("end_id")),
      rename, partitions = 1)
```

**Comment by moxious** [comment]
Hi - I’m writing with good news.  Over the last few months, we’ve been investing a lot of effort in bringing the neo4j spark connector up to newer versions of spark, supporting the Datasource API, and enabling multi-language support for things like Python and R.

Today we’re putting out a pre-release of this new version of the spark connector.

You can download a build here:
https://github.com/neo4j-contrib/neo4j-spark-connector/releases/tag/4.0.0-pre1

And see the documentation here:
http://5.9.211.195:8000/neo4j-spark-docs/1.0.0/

We’re in the process of updating our guidance & documentation on spark.  We’re sorry, but we are unable fix issues & carry forward support of the earlier version of the connector, based on a different API - which went up to and included version 2.4.5-M2.  

Moving forward, we will be focusing our efforts on the new approach above.

As a result, we’re closing the remaining issues attached to the old version, but we would very much appreciate feedback & issues raised against the new 4.0 series of the connector, which you’ll find on the 4.0 branch of the repo.  The connector is going to stay open source, and our current issue list & github planning is in the open and will stay that way.  

We hope you find it a big improvement over the previous version,  and would welcome collaboration around this!
